pkg/okteto/sse.go

```diff
@@ -0,0 +1,43 @@
+// Copyright 2022 The Okteto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package okteto
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/okteto/okteto/pkg/sse"
```

Comment:
I think we can avoid depending on the `github.com/okteto/okteto/pkg/sse` by injecting the streamer. We can define it in place. Then the generic `sse` streamer will happen to implement it.


```golang
type Streamer interface {
  Stream(ctx context.Context, c *http.Client, url string) error
}
type SSEClient struct {
	client *http.Client
        streamer Streamer
}

func NewSSEClient(c *http.Client, s Streamer) *SSEClient {
  return &SSEClient{c, s}
}
```

I know this is not a complete example because sseClient is not public but it's more about the pattern.

The annoying part of this is having to always pass these parameters but I always think that the separation is worth it.

Another approach I'm leaning towards now is defining `Options`. It's a little more involved:

```golang

// okteto package
type Streamer interface {
  Stream(ctx context.Context, c *http.Client, url string) error
}
type SSEClient struct {
	client *http.Client
        streamer Streamer
}

type Option func(*SSEClient)

func NewSSEClient(opts Options) *SSEClient {
  sse := &SSEClient {
    client: http.DefaultClient, // default
    streamer: devNullOrLoggerStreamer, // default
  }
  for _, fn := range opts {
    sse = fn(sse)
  }
  return sse
}

func WithStreamer(s Streamer) Option {
  return func(c *SSEClient){
    c.streamer = s
  }
}

func WithHTTPClient(httpC *http.Client) Option {
  return func(c *SSEClient){
    c.client = httpC
  }
}

// client
func main() {
  c := okteto.NewSSEClient(
    okteto.WithStreamer(sse.Streamer),
    okteto.WithClient(someCustomClient),
  )
}

```

Reply:
FWIW, not saying we should change what we have just sharing some ideas

Reply:
> I think we can avoid depending on the `github.com/okteto/okteto/pkg/sse` by injecting the streamer. We can define it in place. Then the generic `sse` streamer will happen to implement it.
> 
> ```go
> type Streamer interface {
>   Stream(ctx context.Context, c *http.Client, url string) error
> }
> type SSEClient struct {
> 	client *http.Client
>         streamer Streamer
> }
> 
> func NewSSEClient(c *http.Client, s Streamer) *SSEClient {
>   return &SSEClient{c, s}
> }
> ```
> 

@maroshii I am not sure i am understanding this. Do you mean to create a Streamer interface at `package okteto` and move the stream logic of `package sse` to the implementation of this streamer under `okteto` ???

Reply:
No, we can have a `Streamer` interface in the okteto package that defines what a Streamer is in that context. Then, the more generic `sse` package will have the implementation of this Streamer that will satisfy that interface so it can be injected. This will remove the package dependency

Reply:
I know that `sse.Stream` is not a struct but just a function so we would need to change that.

This pattern comes in handy for testing where you can inject any streamer that you want to this function and test that logic in isolation without ever using `sse.Stream`